keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,1009 @@
+  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/kubelet_pods.go#L988
+
+Specifying a `pullPolicy` be supported in the same way as for the existing
+`imagePullPolicy` API.
+
+For registry authentication purposes the same logic will be used as for the
+container image.
+
+#### CRI
+
+The CRI API is already capable of managing container images [via the `ImageService`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L146-L161).
+Those RPCs will be re-used for managing OCI artifacts, while the [`ImageSpec`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L798-L813)
+as well as [`PullImageResponse`](https://github.com/kubernetes/cri-api/blob/3a66d9d/pkg/apis/runtime/v1/api.proto#L1530-L1534)
+will be extended to mount the OCI object to a local path:
+
+```protobuf
+
+// ImageSpec is an internal representation of an image.
+message ImageSpec {
+    // …
+
+    // Indicate that the OCI object should be mounted.
+    bool mount = 20;
+
+    // SELinux label to be used.
+    string mount_label = 21;
+}
+
+message PullImageResponse {
+    // …
+
+    // Absolute local path where the OCI object got mounted.
+    string mountpoint = 2;
+}
+```
+
+This allows to re-use the existing kubelet logic for managing the OCI objects,
+with the caveat that the new `VolumeSource` won't be isolated in a dedicated
+plugin as part of the existing [volume manager](https://github.com/kubernetes/kubernetes/tree/6d0aab2/pkg/kubelet/volumemanager).
+
+The added `mount_label` allow the kubelet to support SELinux contexts.
+
+#### Container Runtimes
+
+Container runtimes need to support the new `mount` field, otherwise the
+feature cannot be used. The kubelet will verify if the returned `mountpoint`
+actually exists on disk to check the feature availability, because Protobuf will
+strip the field in a backwards compatible way for older runtimes. Pods using the
+new `VolumeSource` combined with a not supported container runtime version will
+fail to run on the node.
+
+For security reasons, volume mounts should set the [`noexec`] and `ro`
+(read-only) options by default.
+
+##### SELinux
+
+Traditionally, the container runtime is responsible of applying SELinux labels
+to volume mounts, which are inherited from the `securityContext` of the pod or
+container. Relabeling volume mounts can be time-consuming, especially when there
+are many files on the volume.
+
+If the following criteria are met, then the kubelet will use the `mount_label`
+field in the CRI to apply the right SELinux label to the mount.
+
+- The operating system must support SELinux
+- The Pod must have at least `seLinuxOptions.level` assigned in the
+  `PodSecurityContext` or all volume using containers must have it set in their
+  `SecurityContexts`. Kubernetes will read the default user, role and type from
+  the operating system defaults (typically `system_u`, `system_r` and `container_t`).
```

Comment:
This seems like yet another way of accomplishing - https://github.com/kubernetes/enhancements/issues/1710 but it doesn't fit with existing design.

Reply:
Moving that discussion into https://github.com/kubernetes/enhancements/pull/4642#discussion_r1636806425